src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/TransactionValidator.cs

```diff
@@ -42,6 +68,75 @@ public static void ValidateTransactionBundle(Hl7.Fhir.Model.Bundle bundle)
+        {
+            if (entry.Request.IfNoneExist == null && !entry.Request.Url.Contains("?", StringComparison.InvariantCulture))
+            {
+                if (entry.Request.Method == HTTPVerb.POST)
+                {
+                    return entry.FullUrl;
+                }
+
+                return entry.Request.Url;
+            }
+            else
+            {
+                string resourceType = null;
+                StringValues conditionalQueries;
+                HTTPVerb requestMethod = (HTTPVerb)entry.Request.Method;
+                bool conditionalCreate = requestMethod == HTTPVerb.POST;
+                bool condtionalUpdate = requestMethod == HTTPVerb.PUT || requestMethod == HTTPVerb.DELETE;
+
+                if (condtionalUpdate)
+                {
+                    string[] conditinalUpdateParameters = entry.Request.Url.Split("?");
+                    resourceType = conditinalUpdateParameters[0];
+                    conditionalQueries = conditinalUpdateParameters[1];
+                }
+                else if (conditionalCreate)
+                {
+                    resourceType = entry.Request.Url;
+                    conditionalQueries = entry.Request.IfNoneExist;
+                }
+
+                SearchResultEntry[] matchedResults = await GetExistingResourceId(entry.Request.Url, resourceType, conditionalQueries);
+                int? count = matchedResults?.Length;
+
+                if (count > 1)
+                {
+                    // Multiple matches: The server returns a 412 Precondition Failed error indicating the client's criteria were not selective enough
+                    throw new PreconditionFailedException(string.Format(Api.Resources.ConditionalOperationNotSelectiveEnough, conditionalQueries));
+                }
+
+                if (count == 1)
+                {
+                    return entry.Resource.TypeName + "/" + matchedResults[0].Resource.ResourceId;
+                }
+            }
+
+            return string.Empty;
+        }
+
+        public async Task<SearchResultEntry[]> GetExistingResourceId(string requestUrl, string resourceType, StringValues conditionalQueries)
+        {
+            if (string.IsNullOrEmpty(resourceType) || string.IsNullOrEmpty(conditionalQueries))
+            {
+                throw new RequestNotValidException(string.Format(Api.Resources.InvalidConditionalReferenceParameters, requestUrl));
+            }
+
+            Tuple<string, string>[] conditionalParameters = QueryHelpers.ParseQuery(conditionalQueries)
+                              .SelectMany(query => query.Value, (query, value) => Tuple.Create(query.Key, value)).ToArray();
+
+            var searchResourceRequest = new SearchResourceRequest(resourceType, conditionalParameters);
+
+            return await Search(searchResourceRequest.ResourceType, searchResourceRequest.Queries, CancellationToken.None);
```

Comment:
Can CancellationToken be passed through the stack? This might be particularly useful for a bundle as they could take some time